File: apps/scala/design-pattern/test/FactoryTest.scala
package designpattern.factory

class FactorySuite extends munit.FunSuite:
  test("PondFactory がカエルと藻を生成する") {
    val animal = PondFactory.createAnimal("カエル太郎")
    val plant = PondFactory.createPlant("池の藻")

    assert(animal.isInstanceOf[Frog])
    assertEquals(animal.name, "カエル太郎")
    assertEquals(animal.speak, "ケロケロ")

    assert(plant.isInstanceOf[Algae])
    assertEquals(plant.name, "池の藻")
  }

  test("JungleFactory がトラとスイレンを生成する") {
    val animal = JungleFactory.createAnimal("虎之助")
    val plant = JungleFactory.createPlant("森のスイレン")

    assert(animal.isInstanceOf[Tiger])
    assertEquals(animal.name, "虎之助")
    assertEquals(animal.speak, "ガオー")

    assert(plant.isInstanceOf[WaterLily])
  }

  test("ファクトリを切り替えて環境を構築する") {
    val pond = Environment(PondFactory)
    assert(pond.describe.contains("ケロケロ"))

    val jungle = Environment(JungleFactory)
    assert(jungle.describe.contains("ガオー"))
  }

  test("Animal trait の共通プロパティ") {
    val duck = Duck("アヒル太郎")
    assertEquals(duck.legs, 2)
    assertEquals(duck.speak, "ガーガー")
  }

File: apps/csharp/design-pattern/tests/DesignPattern.Tests/Patterns/FactoryTest.cs
using DesignPattern.Patterns.Factory;

namespace DesignPattern.Tests.Patterns;

public class FactoryTest
{
    [Fact]
    public void Duck_SpeaksQuack()
    {
        IAnimal duck = new Duck();

        Assert.Equal("Quack!", duck.Speak());
    }

    [Fact]
    public void DuckPond_CreatesDuckAndWaterLily()
    {
        var pond = new DuckPond();

        Assert.IsType<Duck>(pond.Animal);
        Assert.IsType<WaterLily>(pond.Plant);
    }

    [Fact]
    public void FrogPond_CreatesFrogAndAlgae()
    {
        var pond = new FrogPond();

        Assert.IsType<Frog>(pond.Animal);
        Assert.IsType<Algae>(pond.Plant);
    }

    [Fact]
    public void Pond_Describe_IncludesAnimalAndPlant()
    {
        var pond = new DuckPond();
        var desc = pond.Describe();

        Assert.Contains("Duck", desc);
        Assert.Contains("WaterLily", desc);
        Assert.Contains("Quack!", desc);
    }

    [Fact]
    public void JungleFactory_CreatesTigerAndTree()
    {
        var factory = new JungleFactory();
        var habitat = new Habitat(factory);

        Assert.IsType<Tiger>(habitat.Animal);
        Assert.IsType<Tree>(habitat.Plant);
    }

    [Fact]
    public void PondFactory_CreatesFrogAndAlgae()
    {
        var factory = new PondFactory();
        var habitat = new Habitat(factory);

        Assert.IsType<Frog>(habitat.Animal);
        Assert.IsType<Algae>(habitat.Plant);
    }

    [Fact]
    public void Habitat_Describe_IncludesInfo()
    {
        var habitat = new Habitat(new JungleFactory());
        var desc = habitat.Describe();

        Assert.Contains("Tiger", desc);
        Assert.Contains("Roar!", desc);
        Assert.Contains("Tree grows tall", desc);
    }
}
